finality-aleph/src/sync/substrate/verification/cache.rs

```diff
@@ -128,9 +156,68 @@ where
     /// Prune all sessions with a number smaller than `session_id`
     fn prune(&mut self, session_id: SessionId) {
         self.sessions.retain(|&id, _| id >= session_id);
+        // off by one!
+        // let's retain one more session, just to be safe
+        self.aura_authorities
+            .retain(|&id, _| id.next() >= session_id);
         self.lower_bound = session_id;
     }
 
+    fn try_prune(&mut self, session_id: &SessionId) {
+        if session_id.0
+            >= self
+                .lower_bound
+                .0
+                .saturating_add(self.cache_size.saturated_into())
+        {
+            self.prune(SessionId(
+                session_id
+                    .0
+                    .saturating_sub(self.cache_size.saturated_into())
+                    + 1,
+            ));
+        }
+    }
+
+    /// Returns the list of Aura authorities for a given block number. Updates cache if necessary.
+    /// This method assumes that the queued Aura authorities will indeed become Aura authorities
+    /// in the next session.
+    pub fn get_aura_authorities(
+        &mut self,
+        number: BlockNumber,
+    ) -> Result<&Vec<AuraId>, CacheError> {
+        if number == 0 {
+            return Err(CacheError::NoAuraAutoritiesForGenesis);
+        }
+        let session_id = self.session_info.session_id_from_block_num(number);
+        if session_id < self.lower_bound {
+            return Err(CacheError::SessionTooOld(session_id, self.lower_bound));
+        }
+        let upper_bound = SessionId(
+            self.session_info
+                .session_id_from_block_num(self.finalization_info.finalized_number())
+                .0,
+        );
+        if session_id > upper_bound {
```

Comment:
cant we check also the next session? We should have confirmed authorities for the `n + 1` session if the first block of session `n` is finalized.

Reply:
fixed